### ISSUE TITLE
Grounding DINO is now available in 🤗 Transformers!

### ISSUE THREAD

**Issue by NielsRogge** [IDEA-Research/GroundingDINO#321]
Hi folks!

Grounding DINO is now available in the Transformers library, enabling easy inference in a few lines of code.

Here's how to use it:

```python
from transformers import AutoProcessor, GroundingDinoForObjectDetection
from PIL import Image
import requests

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
text = "a cat."

processor = AutoProcessor.from_pretrained("IDEA-Research/grounding-dino-tiny")
model = GroundingDinoForObjectDetection.from_pretrained("IDEA-Research/grounding-dino-tiny")

inputs = processor(images=image, text=text, return_tensors="pt")
outputs = model(**inputs)

# convert outputs (bounding boxes and class logits) to COCO API
target_sizes = torch.tensor([image.size[::-1]])
results = processor.image_processor.post_process_object_detection(
    outputs, threshold=0.35, target_sizes=target_sizes
)[0]
for score, label, box in zip(results["scores"], results["labels"], results["boxes"]):
    box = [round(i, 2) for i in box.tolist()]
    print(f"Detected {label.item()} with confidence " f"{round(score.item(), 3)} at location {box}")
```

Demo notebook: https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Grounding%20DINO/Inference_with_Grounding_DINO_for_zero_shot_object_detection.ipynb

Checkpoints are on the hub: https://huggingface.co/models?other=grounding-dino

**Comment by NielsRogge** [comment]
Relevant for https://github.com/IDEA-Research/GroundingDINO/issues/88

**Comment by Lycus99** [comment]
Thanks for your work!
I found an issue where the same model produces different results. Take the image of '000000039769.jpg' as an example. The results of the official code are significantly better than those of the transformer library

The results you report are         
**Detected 1 with confidence 0.45 at location [344.8, 23.2, 637.4, 373.8]
Detected 1 with confidence 0.41 at location [11.9, 51.6, 316.6, 472.9]**


My results based on the transformer package
Code: 
from transformers import AutoProcessor, GroundingDinoForObjectDetection
from PIL import Image
import requests
import torch

import matplotlib.pyplot as plt
import matplotlib.patches as patches

image = Image.open('000000039769.jpg')

text = "cat"
device = 'cpu'
processor = AutoProcessor.from_pretrained("IDEA-Research/grounding-dino-tiny")
model = GroundingDinoForObjectDetection.from_pretrained("IDEA-Research/grounding-dino-tiny").to(device)
inputs = processor(images=image, text=text, return_tensors="pt").to(device)
outputs = model(**inputs)

# convert outputs (bounding boxes and class logits) to COCO API
target_sizes = torch.tensor([image.size[::-1]])
results = processor.image_processor.post_process_object_detection(
    outputs, threshold=0.2, target_sizes=target_sizes
)[0]

for score, label, box in zip(results["scores"], results["labels"], results["boxes"]):
    box = [round(i, 2) for i in box.tolist()]
    print(f"Detected {label.item()} with confidence " f"{round(score.item(), 3)} at location {box}")

Results:
**Detected 1 with confidence 0.26 at location [40.29, 72.75, 175.84, 117.19]**

The results based on the official code and colab.

import os

CONFIG_PATH = os.path.join(HOME, "GroundingDINO/groundingdino/config/GroundingDINO_SwinT_OGC.py")
print(CONFIG_PATH, "; exist:", os.path.isfile(CONFIG_PATH))
WEIGHTS_NAME = "groundingdino_swint_ogc.pth"
WEIGHTS_PATH = os.path.join(HOME, "weights", WEIGHTS_NAME)
print(WEIGHTS_PATH, "; exist:", os.path.isfile(WEIGHTS_PATH))

from groundingdino.util.inference import load_model, load_image, predict, annotate

model = load_model(CONFIG_PATH, WEIGHTS_PATH)
import os
import supervision as sv

IMAGE_NAME = "000000039769.jpg"
IMAGE_PATH = os.path.join(HOME, "data", IMAGE_NAME)

TEXT_PROMPT = "cat"
BOX_TRESHOLD = 0.35
TEXT_TRESHOLD = 0.25

image_source, image = load_image(IMAGE_PATH)

boxes, logits, phrases = predict(
    model=model, 
    image=image, 
    caption=TEXT_PROMPT, 
    box_threshold=BOX_TRESHOLD, 
    text_threshold=TEXT_TRESHOLD
)

annotated_frame = annotate(image_source=image_source, boxes=boxes, logits=logits, phrases=phrases)

%matplotlib inline  
sv.plot_image(annotated_frame, (16, 16))

![image](https://github.com/IDEA-Research/GroundingDINO/assets/109274751/a073be1f-4c2d-4e58-ae9a-652846ac3e62)

**Comment by NielsRogge** [comment]
Pinging @eduardopach here

**Comment by EduardoPach** [comment]
> Thanks for your work! I found an issue where the same model produces different results. Take the image of '000000039769.jpg' as an example. The results of the official code are significantly better than those of the transformer library
> 
> The results you report are **Detected 1 with confidence 0.45 at location [344.8, 23.2, 637.4, 373.8] Detected 1 with confidence 0.41 at location [11.9, 51.6, 316.6, 472.9]**
> 
> My results based on the transformer package Code: from transformers import AutoProcessor, GroundingDinoForObjectDetection from PIL import Image import requests import torch
> 
> import matplotlib.pyplot as plt import matplotlib.patches as patches
> 
> image = Image.open('000000039769.jpg')
> 
> text = "cat" device = 'cpu' processor = AutoProcessor.from_pretrained("IDEA-Research/grounding-dino-tiny") model = GroundingDinoForObjectDetection.from_pretrained("IDEA-Research/grounding-dino-tiny").to(device) inputs = processor(images=image, text=text, return_tensors="pt").to(device) outputs = model(**inputs)
> 
> target_sizes = torch.tensor([image.size[::-1]]) results = processor.image_processor.post_process_object_detection( outputs, threshold=0.2, target_sizes=target_sizes )[0]
> 
> for score, label, box in zip(results["scores"], results["labels"], results["boxes"]): box = [round(i, 2) for i in box.tolist()] print(f"Detected {label.item()} with confidence " f"{round(score.item(), 3)} at location {box}")
> 
> Results: **Detected 1 with confidence 0.26 at location [40.29, 72.75, 175.84, 117.19]**
> 
> The results based on the official code and colab.
> 
> import os
> 
> CONFIG_PATH = os.path.join(HOME, "GroundingDINO/groundingdino/config/GroundingDINO_SwinT_OGC.py") print(CONFIG_PATH, "; exist:", os.path.isfile(CONFIG_PATH)) WEIGHTS_NAME = "groundingdino_swint_ogc.pth" WEIGHTS_PATH = os.path.join(HOME, "weights", WEIGHTS_NAME) print(WEIGHTS_PATH, "; exist:", os.path.isfile(WEIGHTS_PATH))
> 
> from groundingdino.util.inference import load_model, load_image, predict, annotate
> 
> model = load_model(CONFIG_PATH, WEIGHTS_PATH) import os import supervision as sv
> 
> IMAGE_NAME = "000000039769.jpg" IMAGE_PATH = os.path.join(HOME, "data", IMAGE_NAME)
> 
> TEXT_PROMPT = "cat" BOX_TRESHOLD = 0.35 TEXT_TRESHOLD = 0.25
> 
> image_source, image = load_image(IMAGE_PATH)
> 
> boxes, logits, phrases = predict( model=model, image=image, caption=TEXT_PROMPT, box_threshold=BOX_TRESHOLD, text_threshold=TEXT_TRESHOLD )
> 
> annotated_frame = annotate(image_source=image_source, boxes=boxes, logits=logits, phrases=phrases)
> 
> %matplotlib inline sv.plot_image(annotated_frame, (16, 16))
> 
> ![image](https://private-user-images.githubusercontent.com/109274751/325882864-a073be1f-4c2d-4e58-ae9a-652846ac3e62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQxMjQzNDQsIm5iZiI6MTcxNDEyNDA0NCwicGF0aCI6Ii8xMDkyNzQ3NTEvMzI1ODgyODY0LWEwNzNiZTFmLTRjMmQtNGU1OC1hZTlhLTY1Mjg0NmFjM2U2Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNlQwOTM0MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yODBkMzczODczM2E1YmI5ZTE3ZWVkNTlhYTM1NjEyMDBiYWJlMmFlNTUwZjFkMjkwZTE2MTZiM2M5ZGIxZTRlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.kMH6Htr0hf5cemd-odU1Vif186kbAscd1OYYMZZ2Tqo)

Actually, the output is the same :D, there's only one catch in your example.

The original implementation takes your text prompt and adds a `.` at the end so what is happening is that for the `transformers` example, you're passing `"cat"` and in the original, you're passing `"cat."` and this causes the difference you're seeing

**Comment by MinGiSa** [comment]
it's possible to train?

**Comment by DAAworld** [comment]
Think you!

**Comment by EduardoPach** [comment]
> it's possible to train?

In theory, you can, it wasn't extensively tested, but if you find any problems open an issue in the [transformers](https://github.com/huggingface/transformers) repo and tag me there

**Comment by XingWang1234** [comment]
@NielsRogge 
Appreciate the great work.

is there any caching mechanism inside Transformers implementation? I used the latest Transformers=4.45.2
when I load the model into GPU memory, total consumed GPU memory is 1318MiB
I only changed the input image, did not change the input text.
after the 1st inference, total consumed GPU memory is 2766MiB
after the 2nd inference, total consumed GPU memory is 3424MiB
after the 3rd inference, total consumed GPU memory is 3424MiB
after the 4th inference, total consumed GPU memory is 3424MiB
after the 5th inference, total consumed GPU memory is 3424MiB
why do the 1st inference and 2nd inference increase GPU memory so much, I wonder if there is some caching mechanism inside the codes?

when I used torch.cuda.empty_cache() to release GPU memory after each inference is finished.
after the 1st inference, total consumed GPU memory is 1516MiB
after the 2nd inference, total consumed GPU memory is 1552MiB
after the 3rd inference, total consumed GPU memory is 1530MiB
after the 4th inference, total consumed GPU memory is 1516MiB
after the 5th inference, total consumed GPU memory is 1542MiB

https://huggingface.co/IDEA-Research/grounding-dino-tiny/discussions/4

Thanks a lot.

**Comment by yaoliliu** [comment]
Hello! I'm just wondering how to handle multi-images in one time, since my code below does not seem to work:
``` python
from transformers import AutoProcessor, GroundingDinoForObjectDetection
from PIL import Image
import requests
import torch
from typing import List
import os
    
def crop_objects(imgs: List, labels: List):
    processor = AutoProcessor.from_pretrained("IDEA-Research/grounding-dino-tiny")
    model = GroundingDinoForObjectDetection.from_pretrained("IDEA-Research/grounding-dino-tiny")
    
    inputs = processor(images=imgs, text=labels, return_tensors="pt")
    outputs = model(**inputs)

    # convert outputs (bounding boxes and class logits) to COCO API
    target_sizes = torch.tensor([img.size[::-1] for img in imgs])
    results = processor.image_processor.post_process_object_detection(
        outputs, threshold=0.35, target_sizes=target_sizes
    )
    cropped_imgs = []
    for res, img in zip(results, imgs):
        for score, label, box in zip(res["scores"], res["labels"], res["boxes"]):
            box = [round(i, 2) for i in box.tolist()]
            print(f"Detected {labels[label.item()]} with confidence " f"{round(score.item(), 3)} at location {box}")
            box = list(map(int, box))
            cropped_img = img.crop(box)
            cropped_imgs.append(cropped_img)
    return cropped_imgs

# for test
if __name__ == '__main__':
    imgs = []
    print(os.getcwd())
    for root, dirs, files in os.walk('./data/predicate/woman_hold_baby/0'):
        for file in files:
            if file.endswith('.jpg'):
                img = Image.open(os.path.join(root,file))
                imgs.append(img)
    labels = [['a man', 'a woman', 'a baby'] for _ in range(len(imgs))]
    cropped_imgs = crop_objects(imgs, labels)
    for i, img in enumerate(cropped_imgs):
        img.save(f'./{i}.jpg')
```

**Comment by NielsRogge** [comment]
We just standardized all zero-shot object detectors in Transformers to allow this! https://www.linkedin.com/posts/pavel-iakubovskii_zero-shot-object-detection-made-easy-with-activity-7286062559349116930-vpcE?utm_source=share&utm_medium=member_ios
